strings/boyer_moore_algorithm.cpp

```diff
@@ -0,0 +1,115 @@
+/**
+ * \file
+ * \brief The [Boyer-Moore-Algorithm]
+ * (https://en.wikipedia.org/wiki/Boyer%E2%80%93Moore_string-search_algorithm)
+ * for finding a pattern within a piece of text with complexity O(n / m)
+ *
+ * 1. Boyer Moore is a combination of the following two approaches.
+ *     -> Bad Character Heuristic
+ *     -> Good Suffix Heuristic
+ * Both of the above heuristics can also be used independently to search a
+ * pattern in a text.
+ * 2. The idea of bad character heuristic is simple. The character of the text 
+ * which doesn’t match with the current character of the pattern is called the Bad 
+ * Character. Upon mismatch, we shift the pattern until – 
+ *     -> he mismatch becomes a match 
+ *     -> Pattern P moves past the mismatched character.
+ */
+
+#include <iostream> /// for IO operations
+#include<cassert> // for assert
+#include<vector> // for std::vector
+#ifdef _MSC_VER
+#include <string>  /// use this for MS Visual C++
+#else
+#include <cstring> // for string
+#endif
+
+
+constexpr int BAD = 256;
+
+/**
+ * Filling Bad Char Heuristic
+ * \param[in] badChar array 
+ * \param[in] m size of pattern string
+ * \param[in] pat string pattern to search
+ * void function fill badChar array
+ */
+
+void fillBadChar(int badChar[], int m, std::string pat) {
+    // First fill it with -1
+    for (int i = 0; i < BAD; i++) badChar[i] = -1;
+
+    // Fix the reuqired values
+    for (int i = 0; i < m; i++) badChar[(int)pat[i]] = i;
+}
+
+/**
+ * @brief Boyer Moore algorithm to find a pattern in a text.
+ * \param[in] text text in which to search
+ * \param[in] pattern string pattern to search
+ * @returns vector of indexes where found
+ */
+
+std::vector<int> boyer_moore(std::string txt, std::string pat) {
+    int n = txt.length();  
+    int m = pat.length();  
```

Comment:
Please rename n, m I really can not infer what it's used for. Name according to the use.